### PR TITLE
refactor(query): split fuse source to read data and deserialize

### PR DESCRIPTION
**Pull request by zhang2014**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

refactor(query): split fuse source to read data and deserialize

Closes #issue

### CODE REVIEW

**Comment by sandflee** [comment]
@zhang2014 , why we are doing this feature? should hive adaptive related logic?

**Comment by BohuTANG** [comment]
@sandflee 

This PR tries to make read data and deserialize data into different transforms and let them more parallel.

**Comment by zhang2014** [comment]
> @zhang2014 , why we are doing this feature? should hive adaptive related logic?

Since hive depends on the metadata of the parquet, there may be no way to split it.

File: src/query/sql/src/planner/optimizer/heuristic/prewhere_optimization.rs [hunk 1]
```diff
@@ -11,133 +11,133 @@
...
-            let children = s_expr
-                .children()
-                .iter()
-                .map(|expr| self.prewhere_optimize(expr.clone()))
-                .collect::<Result<Vec<_>>>()?;
-            Ok(SExpr::create(rel_op.clone(), children, None, None))
-        }
-    }
-}
+//
```
Comment (sundy-li) [inline src/query/sql/src/planner/optimizer/heuristic/prewhere_optimization.rs:14]: `prewhere` may be useful for other engines like Parquet engine.

We can turn off the prewhere in catalog:

```
fn support_prewhere(&self) -> bool {
        false
}
```

For fusetable.

File: src/query/service/src/sessions/query_ctx.rs [hunk 2]
```diff
@@ -232,6 +232,22 @@ impl TableContext for QueryContext {
         self.partition_queue.write().pop_front()
     }
 
+    fn try_get_parts(&self, num: usize) -> Vec<PartInfoPtr> {
+        let mut res = Vec::with_capacity(num);
+        let mut partition_queue = self.partition_queue.write();
+
+        for _index in 0..num {
+            if let Some(part) = partition_queue.pop_front() {
```
Comment (BohuTANG) [inline src/query/service/src/sessions/query_ctx.rs:240]:
```suggestion
            if let Some(part) = partition_queue.pop_front() {
                res.push(part);
            } else {
            break;
            }
```

File: src/query/storages/fuse/src/operations/read/native_data_source_reader.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        }
+
+        Ok(Event::Async)
+    }
+
+    async fn async_process(&mut self) -> Result<()> {
+        let mut parts = Vec::with_capacity(self.batch_size);
+        let mut chunks = Vec::with_capacity(self.batch_size);
+
+        for _index in 0..self.batch_size {
```
Comment (BohuTANG) [inline src/query/storages/fuse/src/operations/read/native_data_source_reader.rs:133]: let parts = self.ctx.try_get_parts(self.batch_size)?

**Review by BohuTANG — APPROVED** [review]

File: src/query/storages/fuse/src/operations/read/native_data_source_reader.rs (hunk 3 above)
Comment (BohuTANG) [reply src/query/storages/fuse/src/operations/read/native_data_source_reader.rs:133]: Else LGTM :/

**Pull request merged**